### DISCUSSION TITLE
[Extending] Obtaining instance ivar values by key?

### DISCUSSION THREAD

**Discussion by STREGAsGate** [marcobambini/gravity#398] [Q&A] [answered]
Hey,

I'm writing a Swift package for Gravity and I'm having trouble figuring out how to get ivar values by key from an Instance.

I see this commented out [Edit] this was for a debug feature, the ivar names are stored in htable.
https://github.com/marcobambini/gravity/blob/a135732f5b0cff4d194a3a4c54f7c02f1700a379/src/shared/gravity_value.h#L382

and I see this function to set values, but the key argument doesn't appear to be used for anything.
[Edit] this function is for expanding the ivar array and getting back it's count.
https://github.com/marcobambini/gravity/blob/f9e95a41c18ed995f2c7cee7498c1a2313427c08/src/shared/gravity_value.c#L211
Is it possible to retrieve an ivar value by it's key, or only by index?

**Comment by STREGAsGate** [comment]
To get an existing instance value:
1. Use `gravity_hash_lookup` or `gravity_hash_lookup_cstring` to get a `gravity_value_t`
2. Cast the `gravity_value_t.p` to `gravity_closure_t`
3. Get the index at `gravity_closure_t.f.index`
4. Use the index to return `instance.ivars[index]` which is the value for the key.

[Edit] the closure is for getter and setters which will need to be handled separately in addition to the stored retrieval above.

**Comment by STREGAsGate** [comment]
This seems to be the basic approach to assign a new stored ivar to a class.

```swift
let index = gravity_class_add_ivar(gravityClass, name)
let function = gravity_function_new_special(gravity.vm, nil, UInt16(index), nil, nil)
let closure = gravity_closure_new(gravity.vm, function)
gravity_class_bind(gravityClass, name, gravity_value_from_object(closure))
```

I'm still having an issue where the closure.f pointer gets corrupted or nulled at some point though.

**Comment by marcobambini** [comment, marked as answer]
Hello @STREGAsGate, can you please describe me what are you trying to do?

**Reply by STREGAsGate** [reply]
Sure!
I'm working on a cross platform Swift Package that makes using gravity a 1 liner for Swift projects, [here](https://github.com/STREGAsGate/Gravity).

**TLDR Question:**
How do I add instance variables to a `gravity_class_t` using the C API?

**Long Question:**
I'm trying to allow `Swift` to describe a `Gravity Class`.
The [documentation](https://marcobambini.github.io/gravity/#/extending?id=extending-gravity) shows how to add a function and that's easy enough and I was able to do that no problem.

To add an instance variable, I currently have [this](https://github.com/STREGAsGate/Gravity/blob/fd8c46abeb80e622ce022b409b0a46c5aa56e6a5/Sources/Gravity/GravityClass.swift#L35-L46) from browsing the Gravity source:
```swift
let index = gravity_class_add_ivar(gravityClass, name)
let function = gravity_function_new_special(nil, nil, UInt16(index), nil, nil)
let closure = gravity_closure_new(gravity.vm, function)
gravity_class_bind(gravityClass, name, gravity_value_from_object(closure))
```

However when attempting to retrieve the value from an instance, [here](https://github.com/STREGAsGate/Gravity/blob/fd8c46abeb80e622ce022b409b0a46c5aa56e6a5/Sources/Gravity/GravityInstance.swift#L35-L46), the `closure.f` is no longer referencing the function.
<sub>This value retrieval works fine for instances that Gravity creates from a script.</sub>

I wrote a [test](https://github.com/STREGAsGate/Gravity/blob/4469652c814a147400789b8991637b3f606c785e/Sources/GravityTests/GravityExtendingTests.swift#L69-L84) with a full example of the failure and how I'm using it.
```swift
func testHostCreateClientClass() throws {        
    let theThingClass = gravity.createClass(named: "TheThing")
    theThingClass.addInstanceVariable(named: "myVar")
        
    try gravity.compile("extern class TheThing; var theThing = TheThing(); func main() {}")
    try gravity.runMain()
        
    let instance = try gravity.getInstance("theThing")
    
    let gValue = gravity.createValue(randomValue)
    instance.setValue(gValue, named: "myVar")
    
    let result = instance.getValue("myVar")
        
    XCTAssertEqual(result.getInt(), randomValue)
}
```

**Reply by STREGAsGate** [reply]
@marcobambini
Fixed.
I thought I was missing a step here, but my issue was from the use of `gravity_closure_free` in an unrelated location earlier on.

**Comment by marcobambini** [comment]
Hello @STREGAsGate, good to know that you fixed the issue.

Please take a look at the Gravity <-> ObjC bridge, I think it contains a lot of valuable code to you:
https://github.com/marcobambini/gravity/blob/master/binding/GravityObjC/ObjC/
